File: app/src/main/java/com/vaiinilla/app/ui/assistant/AssistantLocalReplies.kt
package com.vaiinilla.app.ui.assistant

import com.vaiinilla.app.domain.model.Product
import com.vaiinilla.app.ui.components.moneyLabel
import java.math.BigDecimal

object AssistantLocalReplies {
    fun reply(
        userMessage: String,
        products: List<Product>,
    ): String {
        val normalized = userMessage.trim().lowercase()
        return when {
            normalized.contains("gluten") -> glutenReply(products)
            normalized.contains("ligero") || normalized.contains("fresco") -> lightReply(products)
            normalized.contains("recomienda") -> recommendReply(products)
            else -> recommendReply(products)
        }
    }

    private fun glutenReply(products: List<Product>): String {
        val safe =
            products.filter { product ->
                val allergens = product.allergens.lowercase()
                !allergens.contains("gluten") && !allergens.contains("trigo")
            }
        return if (safe.isEmpty()) {
            "Prueba frutas o jamaica; confirma alérgenos en cocina."
        } else {
            val names = safe.take(2).joinToString(" o ") { it.name }
            "Sin gluten detectable: $names. Confirma alérgenos en cocina antes de pedir."
        }
    }

    private fun lightReply(products: List<Product>): String {
        val light =
            products.filter { product ->
                val name = product.name.lowercase()
                name.contains("jamaica") || name.contains("fruta") || name.contains("agua")
            }
        val picks =
            if (light.isNotEmpty()) {
                light.take(2).joinToString(" y ") { "${it.name} (${moneyLabel(it.digitalPrice)})" }
            } else {
                "Agua de jamaica o fruta de temporada"
            }
        return "Para algo ligero y fresco te sugiero $picks. Perfecto si quieres acompañar sin sentirte pesado."
    }

    private fun recommendReply(products: List<Product>): String {
        val burrito =
            products.firstOrNull { it.name.contains("burrito", ignoreCase = true) }
                ?: products.firstOrNull { it.categoryId == 20 }
                ?: products.firstOrNull()
        return if (burrito != null) {
            "Te recomiendo el ${burrito.name} por ${moneyLabel(burrito.digitalPrice)}. " +
                "Es de los favoritos del menú y llega en unos ${burrito.estimatedTimeMinutes} min."
        } else {
            "Explora el menú de hoy: hay opciones rápidas y llenadoras. ¿Buscas algo específico?"
        }
    }

    fun filterByChip(
        chip: String,
        products: List<Product>,
    ): List<AssistantRecommendation> {
        val available = products.filter(Product::available)
        return when (chip) {
            "Menos de \$60" -> budgetRecommendations(available)
            "Algo ligero" ->
                available
                    .filter {
                        val name = it.name.lowercase()
                        name.contains("jamaica") ||
                            name.contains("fruta") ||
                            name.contains("agua") ||
                            it.estimatedTimeMinutes <= 5
                    }.take(3)
                    .map { it.toRecommendation() }
            "Combo con bebida" -> {
                val food = available.filter { it.categoryId == 20 }.take(2)
                val drink = available.filter { it.categoryId == 10 }.firstOrNull()
                (food + listOfNotNull(drink)).map { it.toRecommendation() }
            }
            else -> defaultRecommendations(available)
        }
    }

    private fun budgetRecommendations(products: List<Product>): List<AssistantRecommendation> {
        val underBudget = products.filter { parsePrice(it.digitalPrice) <= BigDecimal("60") }
        val preferred = listOf("waffle", "torta", "quesadilla")
        val ordered =
            preferred.mapNotNull { keyword ->
                underBudget.firstOrNull { it.name.contains(keyword, ignoreCase = true) }
            }
        val rest = underBudget.filter { product -> ordered.none { it.id == product.id } }
        return (ordered + rest).take(3).map { it.toRecommendation() }
    }

    private fun defaultRecommendations(products: List<Product>): List<AssistantRecommendation> {
        val demoKeywords = listOf("torta", "burrito", "quesadilla")
        return demoKeywords
            .mapNotNull { keyword ->
                products.firstOrNull { it.name.contains(keyword, ignoreCase = true) }
            }.map { it.toRecommendation() }
    }

    private fun Product.toRecommendation(): AssistantRecommendation {
        val categoryLabel = if (categoryId == 10) "Bebida" else "Comida"
        val minTime = (estimatedTimeMinutes - 2).coerceAtLeast(1)
        return AssistantRecommendation(
            productId = id,
            name = name,
            meta = "$minTime–$estimatedTimeMinutes min · $categoryLabel",
            price = digitalPrice,
            imageUrl = imageUrl,
        )
    }

    private fun parsePrice(value: String): BigDecimal =
        runCatching {
            BigDecimal(value)
        }.getOrDefault(BigDecimal.ZERO)
}

data class AssistantRecommendation(
    val productId: Int?,
    val name: String,
    val meta: String,
    val price: String,
    val imageUrl: String,
)

File: expo/src/domain/assistant-local-replies.ts
import { parseMoney } from '@/domain/contract-rules';
import { moneyLabel, type Product } from '@/domain/models';

export interface AssistantRecommendation {
  productId: number | null;
  name: string;
  meta: string;
  price: string;
  imageUrl: string;
}

export function assistantReply(userMessage: string, products: Product[]): string {
  const normalized = userMessage.trim().toLowerCase();
  if (normalized.includes('gluten')) {
    return glutenReply(products);
  }
  if (normalized.includes('ligero') || normalized.includes('fresco')) {
    return lightReply(products);
  }
  return recommendReply(products);
}

function glutenReply(products: Product[]): string {
  const safe = products.filter((product) => {
    const allergens = product.allergens.toLowerCase();
    return !allergens.includes('gluten') && !allergens.includes('trigo');
  });
  if (safe.length === 0) {
    return 'Prueba frutas o jamaica; confirma alérgenos en cocina.';
  }
  const names = safe
    .slice(0, 2)
    .map((product) => product.name)
    .join(' o ');
  return `Sin gluten detectable: ${names}. Confirma alérgenos en cocina antes de pedir.`;
}

function lightReply(products: Product[]): string {
  const light = products.filter((product) => {
    const name = product.name.toLowerCase();
    return name.includes('jamaica') || name.includes('fruta') || name.includes('agua');
  });
  const picks =
    light.length > 0
      ? light
          .slice(0, 2)
          .map((product) => `${product.name} (${moneyLabel(product.digitalPrice)})`)
          .join(' y ')
      : 'Agua de jamaica o fruta de temporada';
  return `Para algo ligero y fresco te sugiero ${picks}. Perfecto si quieres acompañar sin sentirte pesado.`;
}

function recommendReply(products: Product[]): string {
  const burrito =
    products.find((product) => product.name.toLowerCase().includes('burrito')) ??
    products.find((product) => product.categoryId === 20) ??
    products[0];
  if (!burrito) {
    return 'Explora el menú de hoy: hay opciones rápidas y llenadoras. ¿Buscas algo específico?';
  }
  return `Te recomiendo el ${burrito.name} por ${moneyLabel(burrito.digitalPrice)}. Es de los favoritos del menú y llega en unos ${burrito.estimatedTimeMinutes} min.`;
}

export function filterByChip(chip: string, products: Product[]): AssistantRecommendation[] {
  const available = products.filter((product) => product.available);
  switch (chip) {
    case 'Menos de $60':
      return budgetRecommendations(available);
    case 'Algo ligero':
      return available
        .filter((product) => {
          const name = product.name.toLowerCase();
          return (
            name.includes('jamaica') ||
            name.includes('fruta') ||
            name.includes('agua') ||
            product.estimatedTimeMinutes <= 5
          );
        })
        .slice(0, 3)
        .map(toRecommendation);
    case 'Combo con bebida': {
      const food = available.filter((product) => product.categoryId === 20).slice(0, 2);
      const drink = available.find((product) => product.categoryId === 10);
      return [...food, ...(drink ? [drink] : [])].map(toRecommendation);
    }
    default:
      return defaultRecommendations(available);
  }
}

function budgetRecommendations(products: Product[]): AssistantRecommendation[] {
  const underBudget = products.filter((product) => parseMoney(product.digitalPrice).lte(60));
  const preferred = ['waffle', 'torta', 'quesadilla'];
  const ordered = preferred
    .map((keyword) => underBudget.find((product) => product.name.toLowerCase().includes(keyword)))
    .filter((product): product is Product => Boolean(product));
  const rest = underBudget.filter((product) => !ordered.some((item) => item.id === product.id));
  return [...ordered, ...rest].slice(0, 3).map(toRecommendation);
}

function defaultRecommendations(products: Product[]): AssistantRecommendation[] {
  const demoKeywords = ['torta', 'burrito', 'quesadilla'];
  return demoKeywords
    .map((keyword) => products.find((product) => product.name.toLowerCase().includes(keyword)))
    .filter((product): product is Product => Boolean(product))
    .map(toRecommendation);
}

function toRecommendation(product: Product): AssistantRecommendation {
  const categoryLabel = product.categoryId === 10 ? 'Bebida' : 'Comida';
  const minTime = Math.max(1, product.estimatedTimeMinutes - 2);
  return {
    productId: product.id,
    name: product.name,
    meta: `${minTime}–${product.estimatedTimeMinutes} min · ${categoryLabel}`,
    price: product.digitalPrice,
    imageUrl: product.imageUrl,
  };
}
